server/src/main/java/org/opensearch/action/search/AbstractSearchAsyncAction.java

```diff
@@ -462,7 +463,10 @@ private void onRequestEnd(SearchRequestContext searchRequestContext) {
     }
 
     private void executePhase(SearchPhase phase) {
-        final Span phaseSpan = tracer.startSpan(SpanCreationContext.server().name("[phase/" + phase.getName() + "]"));
+        Span phaseSpan = NoopSpan.INSTANCE;
+        if ((phase instanceof WrappingSearchAsyncActionPhase) == false) {
```

Comment:
1. Please use `currentPhaseHasLifecycle` 
2. I think we could start span here but end it up in this block of the code as well

Reply:
1. `currentPhaseHasLifecycle` is not defined until onPhaseStart(phase) is called. We would have to switch the order to:
```
private void executePhase(SearchPhase phase) {
        onPhaseStart(phase);
        Span phaseSpan = NoopSpan.INSTANCE;
        if (currentPhaseHasLifecycle) {
            phaseSpan = tracer.startSpan(SpanCreationContext.server().name("[phase/" + phase.getName() + "]"));
        }
        try (final SpanScope scope = tracer.withSpanInScope(phaseSpan)) {
            phase.recordAndRun();
        }
...
```
2. Can you elaborate?

Reply:
The "if" check on L467 is necessary because CanMatch phase will still pass here twice and create 2 spans.

Reply:
```
private void executePhase(SearchPhase phase) {
        Span phaseSpan = phaseSpan = tracer.startSpan(SpanCreationContext.server().name("[phase/" + phase.getName() + "]"));
        try (final SpanScope scope = tracer.withSpanInScope(phaseSpan)) {
            onPhaseStart(phase);
            phase.recordAndRun();
        } catch (Exception e) {
            if (logger.isDebugEnabled()) {
                logger.debug(new ParameterizedMessage("Failed to execute [{}] while moving to [{}] phase", request, phase.getName()), e);
            }

            if (currentPhaseHasLifecycle = false) {
                 span.setError(e);
            }

            onPhaseFailure(phase, "", e);
        } finally {
            if (currentPhaseHasLifecycle = false) {
                 span.endSpan();
            }
        }
    }
```

Reply:
This solves the un-closed span issue but will result in two coordinator query phase spans when there should only be only one. It makes more sense to gate at span creation since we know canMatch causes a certain behavior and can account for that from the get go. Ideally we'd use `currentPhaseHasLifecycle` but shouldn't let its absence prevent us from taking that route.

Reply:
> This solves the un-closed span issue but will result in two coordinator query phase spans when there should only be only one. 

Could you please elaborate on that? Where post-CanMatch phase will still pass here twice and create 2 spans? If this is happening - we need to fix that or completely phase out listener usage since it is unreliable.

Reply:
The listener hook `onPhaseStart` is only called in `executePhase`. We already discussed listener is called twice, so therefore `executePhase` (and span creation) is also.

Right now it is fixed by putting `startSpan` inside the if statement. 
```
Span phaseSpan = NoopSpan.INSTANCE;
if ((phase instanceof WrappingSearchAsyncActionPhase) == false) {
    phaseSpan = tracer.startSpan(SpanCreationContext.server().name("[phase/" + phase.getName() + "]"));
}
```
However in your recommendation above that was not the case. We can improve checking for `currentPhaseHasLifecycle` and `WrappingSearchAsyncActionPhase` but should not abandon listener interface as it provides the most accurate phase start and end points for async search execution.

Reply:
> The listener hook `onPhaseStart` is only called in `executePhase`. We already discussed listener is called twice, so therefore `executePhase` (and span creation) is also.

Wait, this was fixed - listeners should not be called for `WrappingSearchAsyncActionPhase`, how come it is called twice?

Reply:
Yes, that is correct. I should have said "listener **WAS** called twice." I am just using as part of the transitive transitive argument to show that `executePhase()` is also called twice.

Reply:
> Yes, that is correct. I should have said "listener **WAS** called twice." I am just using as part of the transitive argument to show that `executePhase()` is also called twice.

Ok, so this should be an issue:
 - `executePhase()` will be called for `WrappingSearchAsyncActionPhase` and have span created / ended
 -  `executePhase()` will be called for real phase  and have span created / ended in the listener

I don't see the issues here, what I am missing?

Reply:
Won't both of these cases create a "query" phase Span? There should only be one query phase.

Reply:
No, that would be span with the name of the action (that `WrappingSearchAsyncActionPhase` use), so it may have just one more parent but should not be an issue

Reply:
I believe the action passed to `WrappingSearchAsyncActionPhase` is either `SearchDfsQueryThenFetchAsyncAction` or `SearchQueryThenFetchAsyncAction` which would result in two query spans. At least that was the behavior before and I don't see where it's changed.

That is a good point they may be nested. Parent span (from WrappingSearchAsyncActionPhase) may close before child span (from actual query phase) which is awkward.

Reply:
> That is a good point they may be nested. Parent span (from WrappingSearchAsyncActionPhase) may close before child span (from actual query phase) which is awkward.

It could be a bit awkward but this is small price to pay for clean logic